core/src/main/scala/org/apache/spark/CacheRecoveryManager.scala

```diff
@@ -0,0 +1,250 @@
+      case (k, v) => execIdsToShutDown.contains(k)
+    }
+    val freeMemOnRemaining = remainingMemStatus.values.map(_._2).sum
+
+    val alreadyReplicatedBlocks: Map[String, Set[RDDBlockId]] =
+      savedBlocks.filterKeys(execIdsToShutDown).map { case (k, v) => k -> v.toSet }.toMap
+
+    val getSizes = GetSizeOfBlocks(alreadyReplicatedBlocks)
+    val alreadyReplicatedBytes = blockManagerMasterEndpoint.askSync[Map[String, Long]](getSizes)
+
+    val bytesToReplicate: Seq[(String, Long)] =
+      expiringMemStatus.map { case (execId, (maxMem, remainingMem)) =>
+        val alreadyReplicated = alreadyReplicatedBytes.getOrElse(execId, 0L)
+        val usedMem = maxMem - remainingMem
+        val toBeReplicatedMem = usedMem - alreadyReplicated
+        execId -> toBeReplicatedMem
+      }.toSeq.sortBy { case (k, v) => v }
+
+    bytesToReplicate.scan(("start", freeMemOnRemaining)) {
+      case ((_, remainingMem), (execId, mem)) => (execId, remainingMem - mem)
+    }.drop(1)
+      .toMap
+      .mapValues(freeMem => freeMem >= 0)
+  }
+
+  /**
+   * Gets a list of cached blocks from the block manager master and updates.
+   * CacheRecoveryManagerState.
+   *
+   * @param execId an executor Id
+   */
+  private def updateBlockState(execId: String): Unit = {
+    val blocks: mutable.Set[RDDBlockId] = blockManagerMasterEndpoint
+      .askSync[collection.Set[BlockId]](GetCachedBlocks(execId))
+      .flatMap(_.asRDDId)(collection.breakOut)
+
+    blocks --= savedBlocks.getOrElse(execId, new mutable.HashSet)
+    blocksToSave(execId) = mutable.PriorityQueue[RDDBlockId](blocks.toSeq: _*)(Ordering.by(_.rddId))
+    killTimers.getOrElseUpdate(execId, scheduleKill(new KillRunner(execId)))
+  }
+
+  class KillRunner(execId: String) extends Runnable { def run(): Unit = killExecutor(execId) }
+  private def scheduleKill(k: KillRunner) = scheduler.schedule(k, forceKillAfterS, TimeUnit.SECONDS)
+
+  /**
+   * Ask block manager master to replicate one cached block.
+   *
+   * @param execId the executor to replicate a block from
+   * @return false if there are no more blocks or if there is an issue
+   */
+  def replicateFirstBlock(execId: String): (Future[Boolean], Option[RDDBlockId]) = synchronized {
+    logDebug(s"Replicate block on executor $execId.")
+    blocksToSave.get(execId) match {
+      case Some(p) if p.nonEmpty => doReplication(execId, p.dequeue())
+      case _ => (Future.successful(false), None)
+    }
+  }
+
+  private def doReplication(
+      execId: String,
+      blockId: RDDBlockId): (Future[Boolean], Option[RDDBlockId]) = {
+    val savedBlocksForExecutor = savedBlocks.getOrElseUpdate(execId, new mutable.HashSet)
+    savedBlocksForExecutor += blockId
+    val replicateMessage = ReplicateOneBlock(execId, blockId, blocksToSave.keys.toSeq)
+    logTrace(s"Started replicating block $blockId on exec $execId.")
+    val future = blockManagerMasterEndpoint.ask[Boolean](replicateMessage)
+    (future, Some(blockId))
```

Comment:
It's a little weird to just return one of the input parameters. Might be easy to simplify this at the call site.

Reply:
this is gone